components/requestreply/backend_pubsub_marshaler.go

```diff
@@ -0,0 +1,60 @@
+package requestreply
+
+import (
+	"encoding/json"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/pkg/errors"
+)
+
+type BackendPubsubMarshaler[Result any] interface {
+	MarshalReply(params BackendOnCommandProcessedParams[Result]) (*message.Message, error)
+	UnmarshalReply(msg *message.Message) (reply Reply[Result], err error)
+}
+
+const (
+	ErrorMetadataKey    = "_watermill_requestreply_error"
+	HasErrorMetadataKey = "_watermill_requestreply_has_error"
+	ResultMetadataKey   = "_watermill_requestreply_result"
+)
+
+type BackendPubsubJSONMarshaler[Result any] struct{}
+
+func (m BackendPubsubJSONMarshaler[Result]) MarshalReply(
+	params BackendOnCommandProcessedParams[Result],
+) (*message.Message, error) {
+	msg := message.NewMessage(watermill.NewUUID(), nil)
+
+	if params.HandleErr != nil {
+		msg.Metadata.Set(ErrorMetadataKey, params.HandleErr.Error())
+		msg.Metadata.Set(HasErrorMetadataKey, "1")
+	} else {
+		msg.Metadata.Set(HasErrorMetadataKey, "0")
+	}
+
+	b, err := json.Marshal(params.HandlerResult)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot marshal reply")
+	}
+
+	msg.Metadata.Set(ResultMetadataKey, string(b))
```

Comment:
Consider using the payload for the results instead of the metadata.

Reply:
done